packages/machine-extractor/src/MachineExtractResult.ts

```diff
@@ -1469,6 +1513,95 @@ export class MachineExtractResult {
           this.machineCallResult.node.loc!.start.column,
     )!;
 
+    return {
+      range: oldRange,
+      newText: reprinted.slice(
+        machineNode.arguments[0].start!,
+        machineNode.arguments[0].end!,
+      ),
+      deleted: deleted.length ? deleted : undefined,
+    };
+  }
+
+  restore({ deleted }: { deleted: DeletedEntity[] }) {
+    // TODO deduplicate pre and post loop stuff
+
+    // this ain't ideal because Recast mutates the input AST
+    // so there is a risk that modifying multiple machines in a single file would lead to problems
+    // however, we never modify multiple machines based on the same file content so this is somewhat safe
+    // each modification updates the AST, that is printed and the file is re-parsed, so the next modification sees the next AST
+    const ast: RecastFile = recast.parse(this._fileContent, {
+      parser: {
+        // this is a slight hack to defer the work done by `recast.parse`
+        // we don't need to re-parse the file though, as we already have the AST
+        parse: () => this._fileAst,
+      },
+    });
+
+    const recastDefinitionNode = findRecastDefinitionNode(
+      ast,
+      this.machineCallResult.definition!.node,
```

Comment:
I'm guessing all the added not-null assertions on the machine results is because they can, in fact, never be null, and we don't want to use the type system to express that?

Reply:
Ye, kinda - the codebase is scattered with this right now anyway. I have plans for bigger refactors that would avoid this kind of a thing